File: CGALWrapper/Polygon2_EEK.cpp
#include "pch.h"
#include "Util.h"
#include "Polygon2_EEK.h"
#include "Polygon2.h"

void* Polygon2_EEK_Create()
{
	return Util::Create<Polygon2<EEK>::Polygon_2>();
}

void* Polygon2_EEK_CreateFromPoints(Point2d* points, int startIndex, int count)
{
	return Polygon2<EEK>::CreateFromPoints(points, startIndex, count);
}

void Polygon2_EEK_Release(void* ptr)
{
	Util::Release<Polygon2<EEK>::Polygon_2>(ptr);
}

int Polygon2_EEK_Count(void* ptr)
{
	return Polygon2<EEK>::Count(ptr);
}

void* Polygon2_EEK_Copy(void* ptr)
{
	return Polygon2<EEK>::Copy(ptr);
}

void Polygon2_EEK_Clear(void* ptr)
{
	Polygon2<EEK>::Clear(ptr);
}

Point2d Polygon2_EEK_GetPoint(void* ptr, int index)
{
	return Polygon2<EEK>::GetPoint(ptr, index);
}

void Polygon2_EEK_GetPoints(void* ptr, Point2d* points, int startIndex, int count)
{
	Polygon2<EEK>::GetPoints(ptr, points, startIndex, count);
}

void Polygon2_EEK_SetPoint(void* ptr, int index, Point2d point)
{
	Polygon2<EEK>::SetPoint(ptr, index, point);
}

void Polygon2_EEK_SetPoints(void* ptr, Point2d* points, int startIndex, int count)
{
	Polygon2<EEK>::SetPoints(ptr, points, startIndex, count);
}

void Polygon2_EEK_Reverse(void* ptr)
{
	Polygon2<EEK>::Reverse(ptr);
}

BOOL Polygon2_EEK_IsSimple(void* ptr)
{
	auto polygon = (CGAL::Polygon_2<EEK>*)ptr;
	return polygon->is_simple();

	//return Polygon2<EEK>::IsSimple<(ptr);
}

BOOL Polygon2_EEK_IsConvex(void* ptr)
{
	return Polygon2<EEK>::IsConvex(ptr);
}

CGAL::Orientation Polygon2_EEK_Orientation(void* ptr)
{
	return Polygon2<EEK>::Orientation(ptr);
}

CGAL::Oriented_side Polygon2_EEK_OrientedSide(void* ptr, Point2d point)
{
	return Polygon2<EEK>::OrientedSide(ptr, point);
}

double Polygon2_EEK_SignedArea(void* ptr)
{
	return Polygon2<EEK>::SignedArea(ptr);
}

void Polygon2_EEK_Translate(void* ptr, Point2d translation)
{
	Polygon2<EEK>::Translate(ptr, translation);
}

void Polygon2_EEK_Rotate(void* ptr, double rotation)
{
	Polygon2<EEK>::Rotate(ptr, rotation);
}

void Polygon2_EEK_Scale(void* ptr, double scale)
{
	Polygon2<EEK>::Scale(ptr, scale);
}

void Polygon2_EEK_Transform(void* ptr, Point2d translation, double rotation, double scale)
{
	Polygon2<EEK>::Transform(ptr, translation, rotation, scale);
}

BOOL Polygon2_EEK_ContainsPoint(void* ptr, Point2d point, CGAL::Orientation orientation, BOOL inculdeBoundary)
{
	return Polygon2<EEK>::ContainsPoint(ptr, point, orientation, inculdeBoundary);
}

File: CGALDotNet/Polygons/Polygon2_EEK.cs
using System;
using System.Collections;
using System.Collections.Generic;

using CGALDotNet.Geometry;

namespace CGALDotNet.Polygons
{
    public sealed partial class Polygon2_EEK : Polygon2
    {

        public Polygon2_EEK()
        {
            SetPtr(Polygon2_EEK_Create());
        }

        public Polygon2_EEK(Point2d[] points)
        {
            Count = points.Length;
            SetPtr(Polygon2_EEK_CreateFromPoints(points, 0, points.Length));
            Update();
        }

        internal Polygon2_EEK(IntPtr ptr) : base(ptr)
        {
            Count = Polygon2_EEK_Count(Ptr);
            Update();
        }

        public override string ToString()
        {
            Update();
            return string.Format("[Polygon2_EEK: Count={0}, IsSimple={1}, Orientation={2}]", 
                Count, IsSimple, Orientation);
        }

        public Polygon2_EEK Copy()
        {
            CheckPtr();
            var copy = new Polygon2_EEK(Polygon2_EEK_Copy(Ptr));
            copy.IsSimple = IsSimple;
            copy.Orientation = Orientation;
            copy.IsUpdated = true;
            return copy;
        }

        public override void Clear()
        {
            Count = 0;
            CheckPtr();
            Polygon2_EEK_Clear(Ptr);
            IsSimple = false;
            Orientation = CGAL_ORIENTATION.ZERO;
            IsUpdated = true;
        }

        public override Point2d GetPoint(int index)
        {
            CheckPtr();
            ErrorUtil.CheckBounds(index, Count);
            return Polygon2_EEK_GetPoint(Ptr, index);
        }

        public override Point2d GetPointWrapped(int index)
        {
            CheckPtr();
            index = MathUtil.Wrap(index, Count);
            return Polygon2_EEK_GetPoint(Ptr, index);
        }

        public override Point2d GetPointClamped(int index)
        {
            CheckPtr();
            index = MathUtil.Clamp(index, 0, Count - 1);
            return Polygon2_EEK_GetPoint(Ptr, index);
        }

        public override void GetPoints(Point2d[] points, int startIndex = 0)
        {
            CheckPtr();
            ErrorUtil.CheckBounds(points, startIndex, Count);
            Polygon2_EEK_GetPoints(Ptr, points, startIndex, Count);
        }

        public override void SetPoint(int index, Point2d point)
        {
            CheckPtr();
            ErrorUtil.CheckBounds(index, Count);
            Polygon2_EEK_SetPoint(Ptr, index, point);
            IsUpdated = false;
        }

        public override void SetPoints(Point2d[] points, int startIndex = 0)
        {
            CheckPtr();
            ErrorUtil.CheckBounds(points, startIndex, Count);
            Polygon2_EEK_SetPoints(Ptr, points, startIndex, Count);
            IsUpdated = false;
        }

        public override void Reverse()
        {
            CheckPtr();
            Polygon2_EEK_Reverse(Ptr);
            ReverseOrientation();
        }

        public override bool FindIfSimple()
        {
            CheckPtr();
            if (Count < 3)
                return false;
            else
                return Polygon2_EEK_IsSimple(Ptr);
        }

        public override bool FindIfConvex()
        {
            CheckPtr();
            if (IsSimple)
                return Polygon2_EEK_IsConvex(Ptr);
            else
                return false;
        }

        public override CGAL_ORIENTATION FindOrientation()
        {
            CheckPtr();
            if (IsSimple)
                return Polygon2_EEK_Orientation(Ptr);
            else
                return CGAL_ORIENTATION.ZERO;
        }

        public override CGAL_ORIENTED_SIDE OrientedSide(Point2d point)
        {
            CheckPtr();
            if (IsSimple)
                return Polygon2_EEK_OrientedSide(Ptr, point);
            else
                return CGAL_ORIENTED_SIDE.UNDETERMINED;
        }

        public override double FindSignedArea()
        {
            CheckPtr();
            if (IsSimple)
                return Polygon2_EEK_SignedArea(Ptr);
            else
                return 0;
        }

        protected override void ReleasePtr()
        {
            Polygon2_EEK_Release(Ptr);
        }

    }
}
